addons/mail/models/mail_thread.py

```diff
@@ -4598,3 +4598,23 @@ def get_views(self, views, options=None):
         if "form" in res["views"] and isinstance(self.env[self._name], self.env.registry['mail.activity.mixin']):
             res["models"][self._name]["has_activities"] = True
         return res
+
+    @api.model
+    def _get_thread_with_access(self, thread_model, thread_id, token, _hash, pid):
```

Comment:
Seems it belongs to a controller (a simple python inheritance is possible for portal). Does not really belong to model code as it does nothing except checking controller parameters.

Reply:
The main issue with this method in particular is that it will be used in many (almost all) routes eventually, which are spread across different controllers. Moving it to controller will lead to strong dependency/inheritance between them, which is going to be more complex to manage.

Other methods specific to one route are fine to keep in controller as suggested (example: _prepare_post_data) as they will ever only be used in one place.

Reply:
Actually code is a bit weird. It could even be removed, and just call the _check_thread_portal_access in portal ? Or have a generic "check_access" in mail and add the portal specific code in portal through inheritance ? It seems there is no use of _get_thread_with_access once we remove the credentials check (that belongs to controllers), and the context setup (which can be done in controllers also).

Not sure if my comment is clear :D . Globally it seems there is some mismatch of model and controller code, but most of the code here is controller code (and can be extracted into class methods or even tools). It really helps splitting models and controller code, as model holds business code, if possible it should not have to deal with tokens and hash check :) .

Cheers !

Reply:
message_post is just a first step for using standard code in portal, as the most canonical method for starting this work. Others will follow. `_check_thread_portal_access` will need to be used in many routes in future commits.
The idea is to avoid having to override each route individually in portal to add the check.

So just put the check in routes in mail (the check uses ACL by default), and override only once in portal (the check itself, to check token hash).

As stated above, controller methods are really annoying to re-use in other controllers.

I'm not against tools, but can we easily override them? Do you have an example?

Reply:
`_prepare_post_data` is moved back to the controller.